sdk/containerregistry/container-registry/src/bearerTokenChanllengeAuthenticationPolicy.ts

```diff
@@ -0,0 +1,282 @@
+ */
+export function bearerTokenChallengeAuthenticationPolicy(
+  options: BearerTokenAuthenticationPolicyOptions
+): PipelinePolicy {
+  const { credential, scopes, challengeCallbacks } = options;
+  const callbacks = {
+    authenticateRequest: challengeCallbacks?.authenticateRequest ?? defaultAuthenticateRequest,
+    authenticateRequestOnChallenge:
+      challengeCallbacks?.authenticateRequestOnChallenge ?? defaultAuthenticateRequestOnChallenge,
+    // If any of the properties is set to undefined, it will replace the default values.
+    ...challengeCallbacks
+  };
+
+  /**
+   * We will retrieve the challenge only if the response status code was 401,
+   * and if the response contained the header "WWW-Authenticate" with a non-empty value.
+   */
+  function getChallenge(response: PipelineResponse): string | undefined {
+    const challenge = response.headers.get("WWW-Authenticate");
+    if (response.status === 401 && challenge) {
+      return challenge;
+    }
+    return;
+  }
+
+  // This function encapsulates the entire process of reliably retrieving the token
+  // The options are left out of the public API until there's demand to configure this.
+  // Remember to extend `BearerTokenAuthenticationPolicyOptions` with `TokenCyclerOptions`
+  // in order to pass through the `options` object.
+  const cycler = createTokenCycler(credential /* , options */);
+
+  return {
+    name: bearerTokenChallengeAuthenticationPolicyName,
+    /**
+     * If there's no challenge parameter:
+     * - It will try to retrieve the token using the cache, or the credential's getToken.
+     * - Then it will try the next policy with or without the retrieved token.
+     *
+     * It uses the challenge parameters to:
+     * - Skip a first attempt to get the token from the credential if there's no cached token,
+     *   since it expects the token to be retrievable only after the challenge.
+     * - Prepare the outgoing request if the `prepareRequest` method has been provided.
+     * - Send an initial request to receive the challenge if it fails.
+     * - Process a challenge if the response contains it.
+     * - Retrieve a token with the challenge information, then re-send the request.
+     */
+    async sendRequest(request: PipelineRequest, next: SendRequest): Promise<PipelineResponse> {
+      // Allows users to easily set the authorization header.
+      function setAuthorizationHeader(accessToken: AccessToken) {
+        request.headers.set("Authorization", `Bearer ${accessToken.token}`);
+      }
+
+      if (callbacks?.authenticateRequest) {
+        await callbacks.authenticateRequest({
+          scopes,
+          request,
+          previousToken: cycler.cachedToken,
+          getToken: cycler.getToken,
+          setAuthorizationHeader
+        });
+      }
+
+      let response: PipelineResponse;
+      try {
+        response = await next(request);
+      } catch (err) {
+        response = err.response;
```

Comment:
Remember to grab the error here and throw it at the end if no challenge was found: [link](https://github.com/Azure/azure-sdk-for-js/pull/13888/commits/b387796034ff14a65485f792e7280d481c3ad502#diff-de6e9d74e66787a9b6c5b95fa48fb1dcfb8f14daf0272624fe5e15431701bf57R227).

I found that bug yesterday! It's pretty easy to miss.

There's a test here if you wanted to have one: [test link](https://github.com/Azure/azure-sdk-for-js/pull/13888/commits/b387796034ff14a65485f792e7280d481c3ad502#diff-6f3429998213610b2f80b8299c0f9f098fcc745e986c482462406f717b90844dR314).

Reply:
I will copy the new one!